src/logging.rs

```diff
@@ -186,10 +212,77 @@ enum TrafficLightButton {
     Green,
 }
 
+#[derive(Component, Clone, Copy, PartialEq, Debug)]
+enum LevelFilterChip {
+    Error,
+    Warn,
+    Info,
+    Debug,
+    Trace,
+}
+
+#[derive(Event)]
+struct ChipToggle(LevelFilterChip);
+
+#[derive(Component)]
+struct ErrLogLineMarker;
+
+type WithOnlyErrLogLine = (
+    With<ErrLogLineMarker>,
+    Without<WarnLogLineMarker>,
+    Without<InfoLogLineMarker>,
+    Without<DebugLogLineMarker>,
+    Without<TraceLogLineMarker>,
+);
+
+#[derive(Component)]
+struct WarnLogLineMarker;
+
+type WithOnlyWarnLogLine = (
+    Without<ErrLogLineMarker>,
+    With<WarnLogLineMarker>,
+    Without<InfoLogLineMarker>,
+    Without<DebugLogLineMarker>,
+    Without<TraceLogLineMarker>,
+);
+
+#[derive(Component)]
+struct InfoLogLineMarker;
+
+type WithOnlyInfoLogLine = (
+    Without<ErrLogLineMarker>,
+    Without<WarnLogLineMarker>,
+    With<InfoLogLineMarker>,
+    Without<DebugLogLineMarker>,
+    Without<TraceLogLineMarker>,
+);
+
+#[derive(Component)]
+struct DebugLogLineMarker;
+
+type WithOnlyDebugLogLine = (
+    Without<ErrLogLineMarker>,
+    Without<WarnLogLineMarker>,
+    Without<InfoLogLineMarker>,
+    With<DebugLogLineMarker>,
+    Without<TraceLogLineMarker>,
+);
+
+#[derive(Component)]
+struct TraceLogLineMarker;
+
+type WithOnlyTraceLogLine = (
+    Without<ErrLogLineMarker>,
+    Without<WarnLogLineMarker>,
+    Without<InfoLogLineMarker>,
+    Without<DebugLogLineMarker>,
+    With<TraceLogLineMarker>,
+);
+
 #[derive(Component, Clone)]
 struct AutoCheckBox;
 
-fn setup_log_viewer_ui(mut commands: Commands, log_viewer_res: Res<LogViewer>) {
+fn setup_log_viewer_ui(mut commands: Commands, log_viewer_res: Res<LogViewerState>) {
```

Comment:
this is getting pretty convuluted now. please split out the UI part into its own file